src/main/java/servlets/UploadReceiptServlet.java

```diff
@@ -0,0 +1,174 @@
+   * form. After Blobstore handles the parsing, storing, and hosting of the image, the form
+   * data and a URL where the image can be accessed is forwarded to this servlet in a POST
+   * request.
+   */
+  @Override
+  public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    BlobstoreService blobstoreService = BlobstoreServiceFactory.getBlobstoreService();
+    String uploadUrl = blobstoreService.createUploadUrl("/upload-receipt");
+
+    response.setContentType("text/html");
+    response.getWriter().println(uploadUrl);
+  }
+
+  /**
+   * When the user submits the upload form, Blobstore processes the image and then forwards the
+   * request to this servlet, which analyzes the receipt image and inserts information
+   * about the receipt into Datastore.
+   */
+  @Override
+  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    Optional<Entity> receipt = createReceiptEntity(request);
+
+    if (!receipt.isPresent()) {
+      logger.severe("Valid JPEG file was not uploaded.");
+      response.sendRedirect("/");
+      return;
+    }
+
+    // Store the receipt entity in Datastore.
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+    datastore.put(receipt.get());
+
+    // TODO: Redirect to receipt analysis page for MVP
+    response.sendRedirect("/");
+  }
+
+  /**
+   * Creates and returns a receipt entity, which includes the receipt image and
+   * information about the receipt, or null if the user didn't upload a JPEG file.
+   */
+  private Optional<Entity> createReceiptEntity(HttpServletRequest request) {
+    Optional<BlobKey> blobKeyOption = getUploadedBlobKey(request, "receipt-image");
+
+    if (!blobKeyOption.isPresent()) {
+      return Optional.empty();
+    }
+
+    BlobKey blobKey = blobKeyOption.get();
+    String imageUrl = getBlobServingUrl(blobKey);
+    long timestamp = System.currentTimeMillis();
+    String label = request.getParameter("label");
+
+    // TODO: Replace hard-coded values using receipt analysis with Cloud Vision.
+    double price = 5.89;
+    String store = "McDonald's";
+    String rawText = "McDonald’s Restaurant \n Order No. 389 \n "
+        + "Qty Item Total \n 1 Big Mac 3.99 \n 1 M Iced Coffee 1.40 \n"
+        + "Subtotal 5.39 \n Tax 0.50 \n Total 5.89";
+
+    // Create an entity with a kind of Receipt.
+    Entity receipt = new Entity("Receipt");
+    receipt.setProperty("blobKey", blobKey);
+    receipt.setProperty("imageUrl", imageUrl);
+    receipt.setProperty("timestamp", timestamp);
+    receipt.setProperty("label", label);
+    receipt.setProperty("price", price);
+    receipt.setProperty("store", store);
+    receipt.setProperty("rawText", rawText);
+
+    return Optional.of(receipt);
+  }
+
+  /**
+   * Returns a URL that points to the uploaded file, or null if the user didn't upload a JPEG file.
```

Comment:
You appear to be returning an empty Optional rather than null if a JPEG file wasn't uploaded.

Reply:
Maybe we can also display a popup to the user that they didn't select a file

Reply:
Added an error message.